### DISCUSSION TITLE
Upsert with filter does not seem to honor the filter

### DISCUSSION THREAD

**Discussion by musabhusaini** [supabase/supabase#27344] [Questions] [answered]
I am trying to use the client to upsert some values into a table while also applying a filter. This should be possible according to the documentation. However, the filter is not being honored. I want to check if I am doing something wrong or if this is a genuine bug. Here is a simplified scenario:

1. Create a table and insert some values:

``` sql
create table
  public.test (
    id bigint generated by default as identity,
    name text not null,
    status text not null,
    control text null,
    value text null,
    constraint test_pkey primary key (id)
  ) tablespace pg_default;

create unique index unique_name_status on public.test using btree (name, status) tablespace pg_default;

insert into public.test (name, status, control, value)
  values
    ('test1', 'test', '123', 'foo'),
    ('test2', 'dev', '', 'bar');
```

2. Try to use `upsert` to update/insert some values:

``` ts
  const { data } = await supabase
    .from("test")
    .upsert(
      [
        {
          name: "test1",
          status: "test",
          value: "bar",
        },
        {
          name: "test2",
          status: "dev",
          value: "foo",
        },
        {
          name: "test3",
          status: "prod",
          value: "baz",
        },
      ],
      {
        onConflict: "name, status",
      },
    )
    .neq("control", "123")
    .select();
  console.log(data);
```

Expected output:

``` jsonl
{"id": 7, "name": "test2", "status": "dev", "control": "", "value": "foo"}
{"id": 10, "name": "test3", "status": "prod", "control": null, "value": "baz"}
```

Actual output:
``` jsonl
{"id": 6, "name": "test1", "status": "test", "control": "123", "value": "bar"}
{"id": 7, "name": "test2", "status": "dev", "control": "", "value": "foo"}
{"id": 10, "name": "test3", "status": "prod", "control": null, "value": "baz"}
```

An equivalent SQL command brings the expected output:

``` sql
insert into public.test (name, status, value)
  values
    ('test1', 'test', 'bar'),
    ('test2', 'dev', 'foo'),
    ('test3', 'prod', 'baz')
  on conflict (name, status) do update
    set value = excluded.value
    where test.control <> '123'
  returning *;
```

What am I missing?

**Comment by silentworks** [comment]
This is not a bug as such, the issue here is the naming of the property brings confusion to people using the supabase-js library. When doing a normal `.select()` we can do filtering on it because this is doing an actual `SELECT` query with postgrest, however in the case of using `.select()` after a `.insert, .upsert, .delete` it is not making a `SELECT` query from postgrest's perspective. All the `.select()` is doing in this case is doing a `RETURNING *` as pointed out in your SQL query example, any additional filter passed to it is ignored completely. The only sort of filtering you can do is vertical filtering which means you can filter the returned data columns only, so you can do `.select('name, status')`. 

You can read more about this in the Postgrest docs https://postgrest.org/en/v12/references/api/preferences.html#full. You can also take a look at the code for the `.select()` you are using above here https://github.com/supabase/postgrest-js/blob/master/src/PostgrestTransformBuilder.ts#L21-L53 and compare it to that of the normal `.select()` here https://github.com/supabase/postgrest-js/blob/master/src/PostgrestQueryBuilder.ts#L57-L98 you will notice they have two different return types. In this instance too the normal `.select()` makes a `GET` request whilst the `.select()` after an `.insert, .upsert, .delete` is done in a `POST` request.

If this is a feature you would like to see in the library it's probably best to open a issue on the Postgrest repository itself as without the underlying implementation the `supabase-js` library cannot do much other than  the current workaround of making two separate requests to get the results you are after.

```js
  const { data } = await supabase
    .from("test")
    .upsert(
      [
        {
          name: "test1",
          status: "test",
          value: "bar",
        },
        {
          name: "test2",
          status: "dev",
          value: "foo",
        },
        {
          name: "test3",
          status: "prod",
          value: "baz",
        },
      ],
      {
        onConflict: "name, status",
      },
    );

  const { data } = await supabase
    .from("test")
    .select()
    .neq("control", "123");
  console.log(data);
```

**Reply by musabhusaini** [reply]
Thanks for your response. The issue for me has less to do with the result of the `select` and more with the fact that the `upsert` does not take the filter into account. That is, all records are updated instead of only the ones that match the condition. Like I showed, I can achieve the same with raw SQL, and the document seems to indicate that filters on `upsert` should be possible. Is there another way to do this?

**Reply by silentworks** [reply, marked as answer]
Sorry I misinterpreted what you originally asked. So from reading the PostgREST docs https://postgrest.org/en/v12/references/api/tables_views.html#upsert the only filter than can be applied to an `upsert` is `eq` and only when updating a single item based on a primary key. The Supabase JS reference docs needs updating as it is incorrect to mention using filters on `upsert`. 

You can perform the same sort of upset as the SQL example you have above by using a postgres function and calling it via the `.rpc` method of the supabase-js library.

Let me ping @steve-chavez to see if he has any other ideas on how to accomplish this or if there is a solution in the works in PostgREST itself.

**Reply by musabhusaini** [reply]
Updating the docs to reflect this limitation would be appreciated. Perhaps updating the `upsert` interface would also help since the current interface misleadingly allows for placing filters that silently do nothing.

I will likely have to resort to raw SQL queries (through RPC or otherwise), but this seems like a simple enough operation that should be supported by PostgREST and by extension supabase-js.

**Reply by steve-chavez** [reply]
The filters do not enter the ON CONFLICT clause, right now you could do this through RPC.

**Comment by steve-chavez** [comment]
We'll check if it's possible on https://github.com/PostgREST/postgrest/issues/3611.
